README.md

```diff
@@ -22,14 +30,83 @@ dependencies: [
 ]
 ```
 
+## Usage
+
+`BaggageContext` is intended to be used in conjunction with the instrumentation of distributed systems. To make this
+instrumentation work, all parties involved operate on the same `BaggageContext` type. These are the three common
+parties, in no specific order, and guidance on how to use `BaggageContext`:
+
+### Frameworks/Libraries
+
+Developers creating frameworks/libraries (e.g. NIO, gRPC, AsyncHTTPClient, ...) which benefit from being instrumented
+should adopt `BaggageContext` as part of their public API. AsyncHTTPClient for example might accept a context like this:
+
+```swift
+let context = BaggageContext()
+client.get(url: "https://swift.org", context: context)
+```
+
+For more information on where to place this argument and how to name it, take a look at the
+["Context-Passing Guidelines"](#Context-Passing-Guidelines).
+
+Generally speaking, framework developers do not care about the specific contents of the `BaggageContext` but need to
+make sure to feed it through the entire system. At cross-cutting boundaries, e.g. right before sending an HTTP
+request, they'd inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the flip-sight, an
+HTTP server should extract the request headers into a `BaggageContext`. Injecting/extracting is part of the
+`swift-tracing` libraries [and documented in its own repository](https://github.com/slashmo/gsoc-swift-tracing).
+
+### Instrumentation
+
+When implementing instrumentation for cross-cutting tools, `BaggageContext` becomes the way you propagate metadata such
+as trace ids. Because each instrument knows what values might be added to the `BaggageContext` they are the ones
+creating `BaggageContextKey` types dictating the type of value associated with each key added to the context:
+
+```swift
+private enum TraceIDKey: BaggageContextKey {
+  typealias Value = String
+}
```

Comment:
Please do take the extra step to always define:

```swift
extension BaggageContextProtocol {
    var traceID: TraceIDKey.Value? {
        get {
            return self[TraceIDKey.self]
        }
        set {
            self[TraceIDKey.self] = newValue
        }
    }
```

Reply:
The keys are an implementation detail

Reply:
```suggestion
}

extension BaggageContextProtocol {
    var traceID: TraceIDKey.Value? {
        get {
            return self[TraceIDKey.self]
        }
        set {
            self[TraceIDKey.self] = newValue
        }
    }
}
```

Reply:
Yeap, please apply